packages/grainIntegration-merkle/contracts/MerkleRedeem.sol

```diff
@@ -0,0 +1,193 @@
+        for(uint i = 0; i < size; i++) {
+            arr[i] = claimed[_begin + i][_contributor];
+        }
+        return arr;
+    }
+
+    /**
+     * Get the bytes32 hash merkle root for a slice of all valid
+     * distributions. This will error if an index for a non-existent
+     * distribution is passed.
+     */
+    function merkleRoots(
+        uint _begin,
+        uint _end
+    )
+        external
+        view
+        returns (bytes32[] memory)
+    {
+        uint size = 1 + _end - _begin;
+        bytes32[] memory arr = new bytes32[](size);
+        for(uint i = 0; i < size; i++) {
+            arr[i] = weekMerkleRoots[_begin + i];
+        }
+        return arr;
+    }
+
+    /**
+     *  Check the validity of a claim attempt
+     */
+    function verifyClaim(
+        address _liquidityProvider,
+        uint _week,
+        uint _claimedBalance,
+        bytes32[] memory _merkleProof
+    )
+        public
+        view
+        returns (bool valid)
+    {
+        bytes32 leaf = keccak256(abi.encodePacked(_liquidityProvider, _claimedBalance));
+        return MerkleProof.verify(_merkleProof, weekMerkleRoots[_week], leaf);
+    }
+
+    /**
+     * Publish a merkleRoot that contributors an validate against to claim
+     * funds. Currently the _claimTime can be set to any time past or present,
+     * but this may not be the case when ultimately released. Allocations
+     * might need to observe some minimum delay period before claims
+     * can successfully execute against it.
+     */
+    function seedAllocations(
+        uint _week,
+        bytes32 _merkleRoot,
+        uint64 _claimTime
+    )
+        external
+    {
+        require(hasRole(SEEDER_ROLE, msg.sender), "MerkleRedeem: Must have Seeder role to post allocation roots");
+        require(weekMerkleRoots[_week] == bytes32(0), "cannot rewrite merkle root");
+        weekMerkleRoots[_week] = _merkleRoot;
+        claimTimes[_week] = _claimTime;
+    }
+
+    /**
+      * Passing a zero will delay the claim until MAX_UINT64, which is
+      * July 21, 2554
+      */
+    function delayAllocation(uint _week, uint64 _newClaimTime) external {
+        require(hasRole(DELAYER_ROLE, msg.sender), "MerkleRedeem: Must have pauser role to modify allocations");
+        require(weekMerkleRoots[_week] != bytes32(0), "cannot delay a non-existent distribution");
+        // It seems unfair to pause a live distribution. This probably warrants
+        // further discussion, but I'd hate to prevent some people from
+        // rightfully claiming funds and not others.
+        require(block.timestamp < claimTimes[_week], "MerkleRedeem: Can't Delay a Live Distribution");
+
+        if(_newClaimTime == 0) {
+          _newClaimTime = uint64(-1);
+        }
+        // The below `require` might not be necessary, and in fact might hinder
+        // some desired user stories, where the `PAUSER_ROLE` holder has the
+        // authority to reauthorize a distribution.
```

Comment:
+1

Reply:
Waiting to discuss with Maker tomorrow to see what they think